README.md
[![Build Status](https://travis-ci.org/key-amb/perl5-App-Memcached-CLI.svg?branch=master)](https://travis-ci.org/key-amb/perl5-App-Memcached-CLI)
# NAME

**memcached-cli** - Interactive/Batch CLI for Memcached

# SYNOPSIS

Run an interactive CLI:

```
memcached-cli <HOST[:PORT] | /path/to/socket> [OPTIONS]
memcached-cli --addr|-a <HOST[:PORT] | /path/to/socket> [OPTIONS]
memcached-cli [OPTIONS]  # Connect to 127.0.0.1:11211
```

These above turns into interactive mode like below:

```
memcached@SERVER:PORT>
memcached@SERVER:PORT> set foo Foo
OK
memcached@SERVER:PORT> add bar Bar 300
OK
memcached@SERVER:PORT> gets foo bar
Key:foo Value:Foo       Length:3B       Flags:0 Cas:219
Key:bar Value:Bar       Length:3B       Flags:0 Cas:220
memcached@SERVER:PORT> \cd 1  # Alias of 'cachedump'
Key:foo Value:Foo       Length:3B       Expire:2016-03-26 00:30:27      Flags:0 Cas:219
Key:bar Value:Bar       Length:3B       Expire:2016-03-26 09:53:19      Flags:0 Cas:220
Key:baz Value:Baz       Length:3B       Expire:2016-03-26 10:49:05      Flags:123       Cas:221
memcached@SERVER:PORT> delete foo
OK
memcached@SERVER:PORT> flush_all
OK
memcached@SERVER:PORT> \q     # Quit interactive mode
```

Run as batch script:

```
memcached-cli [options] <command> [<args>]
```

Show Help/Manual:

```
# For general usage
memcached-cli -h|--help
memcached-cli --man

# For available commands
memcached@SERVER:PORT> help
memcached@SERVER:PORT> help <command>
```

# DESCRIPTION

This script runs an interactive CLI or batch utility for Memcached.

In interactive mode, it connects to a specified Memcached server and
interactively executes each command you run.

In batch mode, you can execute any command which you can do in interactive mode.

# COMMANDS

NOTE:

A couple of features of following commands derives from
[memcached/memcached-tool](https://github.com/memcached/memcached/blob/master/scripts/memcached-tool)

- **display|\\d**

    Displays slabs statistics.

    This command comes from _memcached/memcached-tool_.

- **stats|\\s** _REGEXP_

    Shows general statistics of memcached server by `stats` command.
    You can filter the parameters of stats by optional _REGEXP_ argument.

    Comes from _memcached/memcached-tool_.

- **settings|config|\\c** _REGEXP_

    Shows memcached server settings by `stats settings` command.
    You can filter the parameters of stats by optional _REGEXP_ argument.

    Comes from _memcached/memcached-tool_, too.

- **cachedump|\\cd** _CLASS_ \[_NUMBER_\]

    Shows detailed information including expiration times of some items in specified
    slab _CLASS_.

    You can specify _NUMBER_ of items to show.
    Without _NUMBER_ option, shows 20 items only by default.

- **detaildump|\\dd**

    Reports statistics about data access using KEY prefix. The default separator for
    prefix is ':'.

    If you have not enabled reporting at Memcached start-up, you can enable it by
    command `detail on`.

    See man **memcached(1)** for details.

- **detail** _MODE_

    Enables or disables stats collection for `stats detail dump` reporting.

    _MODE_ should be either "on" or "off" to enable or to disable.

- **dump\_all**

    Dumps whole data in Memcached server.

    This command comes from _memcached/memcached-tool_.

    Recommended to use in batch mode like bellow:

    ```
    memcached-cli SERVER:PORT dump_all > /path/to/dump.txt
    ```

- **restore\_dump** _FILE_

    Restore data from dump data file created by `dump_all`.

- **randomset|sample** \[_NUMBER_ \[_MAX\_LENGTH_ \[_MIN\_LENGTH_ \[_NAMESPACE_\]\]\]\]

    Generates random sample data and SET all of them.

    By default, it generates 100 data whose length is between 1B and 1000kB with prefix "memcached-cli:sample";

- **get** _KEY1_ \[_KEY2_ ...\]

    Gets items in memcached by specified _KEYs_ and shows their data.

- **gets** _KEY1_ \[_KEY2_ ...\]

    Gets items with _CAS_ data in memcached by specified _KEYs_ and shows their
    data.

- **set** _KEY_ _VALUE_ \[_EXPIRE_ \[_FLAGS_\]\]

    Stores data into memcached by specified _KEY_, _VALUE_ and optional _EXPIRE_
    and _FLAGS_.

- **add** _KEY_ _VALUE_ \[_EXPIRE_ \[_FLAGS_\]\]

    Stores data into memcached by specified _KEY_, _VALUE_ and optional _EXPIRE_
    and _FLAGS_ only when there is NO data with the same _KEY_ in the server.

- **replace** _KEY_ _VALUE_ \[_EXPIRE_ \[_FLAGS_\]\]

    Stores data into memcached by specified _KEY_, _VALUE_ and optional _EXPIRE_
    and _FLAGS_ only when there IS data with the same _KEY_ in the server.

- **append** _KEY_ _VALUE_

    Appends _VALUE_ after existing data in memcached which has specified _KEY_.

- **prepend** _KEY_ _VALUE_

    Puts _VALUE_ before existing data in memcached which has specified _KEY_.

- **cas** _KEY_ _VALUE_ _CAS_ \[_EXPIRE_ \[_FLAGS_\]\]

    Stores data into memcached by specified _KEY_, _VALUE_ and optional _EXPIRE_
    and _FLAGS_ only when _CAS_ of data is not changed from specified _CAS_ value.

- **touch** _KEY_ _EXPIRE_

    Update data expiration time with specified _KEY_ and _EXPIRE_.

- **incr** _KEY_ _VALUE_

    Add numeric _VALUE_ for a data with specified _KEY_.

- **decr** _KEY_ _VALUE_

    Subtract numeric _VALUE_ from a data with specified _KEY_.

- **delete** _KEY_

    Deletes one item in memcached by specified _KEY_.

- **flush\_all|flush** \[_DELAY_\]

    Invalidates all data in memcached by `flush_all` command.

    With _DELAY_ option, invalidation is delayed for specified seconds.

    See official documentation of _memcached_ for details.

- **call** _COMMAND_ \[_ARGS_ ...\]

    Executes any commands given as arguments.

    With this command, you can execute any command against Memcached including what
    is not implemented as normal commands in this program.

    Here are some examples:

    ```
    > call stats conns
    > call slabs reassign 6 10
    > call flush_all
    ```

    LIMITATION:

    Multi-lined queries like `set` are not supported.

- **version**

    Shows memcached server version.

- **quit|exit|\\q**

    Exits program in interactive mode.

- **help|\\h** \[_COMMAND_\]

    Shows available _COMMANDs_ and summary of their usage.

    With optional _COMMAND_ argument, shows detailed information of it.

# OPTIONS

- **-t|--timeout=Int**

    Sets connection timeout. Default is 1 seconds.

- **-d|--debug**

    Shows debug logs.

# SEE ALSO

[App::Memcached::CLI::Main](https://metacpan.org/pod/App::Memcached::CLI::Main),
**memcached(1)**,
[http://www.memcached.org/](http://www.memcached.org/)

# AUTHORS

YASUTAKE Kiyoshi <yasutake.kiyoshi@gmail.com>

# LICENSE

Copyright (C) 2015 YASUTAKE Kiyoshi.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.  That means either (a) the GNU General Public
License or (b) the Artistic License.


Changes
Revision history for Perl extension App-Memcached-CLI

{{$NEXT}}

v0.9.5 2018-03-28T08:49:45Z

    - Fix test: require script by full path (#2)

v0.9.4 2016-03-26T06:04:43Z

    - Bug Fix: in "dump_all" command, set expiration time 0 correctly for data
      which never expires.
    - Feature: implement "restore_dump" command to load data from dump file
      created by "dump_all".

v0.9.3 2016-03-26T05:25:48Z

    - Feature: implement "dump_all" command as "dump" mode in "memcached-tool"
    - Change: remove alias command "dump" as "cachedump" because confusing

v0.9.2 2016-03-26T04:56:42Z

    - Feature: implement "randomset" command to generate random data and SET
      them into Memcached.

v0.9.1 2016-03-26T04:13:00Z

    - Bug Fix: check connection before each command execution to handle the case
      server goes away.

v0.9.0 2016-03-26T03:42:39Z

    - Feature: implement "call" command which executes any command given as
      arguments for Memcached.
    - Improve: set timeout or improve error handling for some commands.

v0.8.0 2016-03-26T00:29:36Z

    - Change:
      - In 'get', 'gets' print in LTSV when multiple keys found
      - Capitalize first letter of every label when print data in LTSV
    - Improve: in 'cachedump' show VALUE, FLAGS, CAS in addition to LENGTH and
      EXPIRE

v0.7.1 2016-03-25T17:16:56Z

    - Feature: implement "incr", "decr" command to add/subtract data value

v0.7.0 2016-03-25T16:45:25Z

    - Feature: implement "touch" command to update EXPIRE of data

v0.6.4 2016-03-25T16:19:04Z

    - Improve: at 'cachedump' show items in more readable LTSV format

v0.6.3 2016-03-24T19:36:15Z

    - Feature: support multiple keys retrieval in "get" and "gets" command

v0.6.2 2016-03-24T18:45:17Z

    - Feature:
      - implement "gets" to fetch data with CAS
      - implement "cas" to set data with CAS check

v0.6.1 2016-03-24T17:41:33Z

    - Feature: implement "add", "replace", "append", "prepend" storage commands
      for Memcached

v0.6.0 2016-03-24T13:31:51Z

    - Feature: add optional argument for "stats" and "settings" to filter
      results

v0.5.4 2016-03-24T11:50:04Z

    - Update documentation. Add commands info into script Pod.

v0.5.3 2016-03-24T07:54:41Z

    - Feature: implement "flush_all" command to invalidate whole data in
      Memcached

v0.5.2 2016-03-24T07:42:15Z

    - Feature: implement "version" command to show server version

v0.5.1 2016-03-24T07:31:50Z

    - Feature: implement "delete" command to delete item in Memcached

v0.5.0 2016-03-24T07:18:28Z

    - Feature: implement "set" command to store item on Memcached
    - Improve: set timeout to read sock not to hang when unexpected response is
      returned by server

v0.4.0 2016-03-24T05:35:52Z

    - Feature: support batch mode with commandline arguments

v0.3.0 2016-03-24T03:40:05Z

    - Feature: implement "detaildump" command for "stats detail dump"
    - Feature: implement "detail" command to toggle stats collection

v0.2.3 2016-03-24T01:26:45Z

    - Change: remove Encode::Guess from dependency. Decode ASCII text only when
      GET item.

v0.2.2 2016-03-23T13:22:39Z

    - Feature: implement "get" command in interactive mode to GET data of
      Memcached

v0.2.1 2016-03-23T08:12:40Z

    - Update Pod

v0.2.0 2016-03-23T05:09:28Z

    - Improve: enable command history by Term::ReadLine support

v0.1.1 2016-03-23T04:25:09Z

    - Improve "help" in Interactive mode

v0.1.0 2016-03-23T01:14:45Z

    - Initial release



lib/App/Memcached/CLI/Help.pm
package App::Memcached::CLI::Help;

use strict;
use warnings;
use 5.008_001;

use version; our $VERSION = 'v0.9.4';

our @COMMANDS_INFO = (
    +{
        command => 'help',
        summary => 'Show help (this)',
    },
    +{
        command => 'version',
        summary => 'Show server version',
    },
    +{
        command => 'quit',
        summary => 'Exit',
    },
    +{
        command => 'display',
        summary => 'Display slabs info',
    },
    +{
        command => 'stats',
        summary => 'Show stats',
        description => <<'EODESC',
Usage:
    > stats [<REGEXP>]
    > stats            # show all stats
    > stats (hit|miss) # filter by REGEXP
EODESC
    },
    +{
        command => 'settings',
        summary => 'Show settings',
        description => <<'EODESC',
Usage:
    > settings [<REGEXP>]
    > settings             # show all stats
    > settings ^(lru|slab) # filter by REGEXP
EODESC
    },
    +{
        command => 'cachedump',
        summary => 'Show cachedump of specified slab',
        description => <<'EODESC',
Usage:
    > cachedump <CLASS> [<NUMBER>]
    > cachedump 1 10
    > cachedump 3     # show default number of items
EODESC
    },
    +{
        command => 'detaildump',
        summary => 'Show detail dump',
        description => <<'EODESC',
Description:
    Report statistics about data access using KEY prefix. The default separator
    for prefix is ':'.
    If you have not enabled reporting at Memcached start-up, run "detail on".
    See man memcached(1) for details.
EODESC
    },
    +{
        command => 'detail',
        summary => 'Enable/Disable detail dump',
        description => <<'EODESC',
Usage:
    > detail on
    > detail off

Description:
    See "\h detaildump"
EODESC
    },
    +{
        command => 'dump_all',
        summary => 'Dump whole server data',
    },
    +{
        command => 'restore_dump',
        summary => 'Restore data from dump',
        description => <<'EODESC',
Usage:
    > restore_dump FILE
EODESC
    },
    +{
        command => 'randomset',
        summary => 'Generate sample data and SET them',
        description => <<'EODESC',
Usage:
    > randomset [<NUMBER> [<MAX_LENGTH> [<MIN_LENGTH> [<NAMESPACE>]]]]
    > randomset                     # generate 100 data
    > randomset 50                  # generate 50 data
    > randomset 50 1024             # data length:   1-1024B
    > randomset 50 1024 256         # data length: 256-1024B
    > randomset 50 1024 256 sample1 # prefix 'sample1:'
EODESC
    },
    +{
        command => 'get',
        summary => 'Get data by KEYs',
        description => <<'EODESC',
Usage:
    > get <KEY1> [<KEY2> ...]
EODESC
    },
    +{
        command => 'gets',
        summary => 'Get data by KEYs with cas',
        description => <<'EODESC',
Usage:
    > gets <KEY1> [<KEY2> ...]
EODESC
    },
    +{
        command => 'set',
        summary => 'Set data by KEY VALUE',
        description => <<'EODESC',
Usage:
    > set <KEY> <VALUE> [<EXPIRE> [<FLAGS>]]
    > set mykey1 MyValue1
    > set mykey2 MyValue2 0     # Never expires. Default
    > set mykey3 MyValue3 120 1
EODESC
    },
    +{
        command => 'add',
        summary => 'Add data by KEY VALUE',
        description => <<'EODESC',
Description:
    Store data if server has no data with the KEY.

Usage:
    > add <KEY> <VALUE> [<EXPIRE> [<FLAGS>]]
    > add mykey1 MyValue1
    > add mykey2 MyValue2 0     # Never expires. Default
    > add mykey3 MyValue3 120 1
EODESC
    },
    +{
        command => 'replace',
        summary => 'Replace data by KEY VALUE',
        description => <<'EODESC',
Description:
    Overwrite data if server has data with the KEY.

Usage:
    > replace <KEY> <VALUE> [<EXPIRE> [<FLAGS>]]
    > replace mykey1 MyValue1
    > replace mykey2 MyValue2 0     # Never expires. Default
    > replace mykey3 MyValue3 120 1
EODESC
    },
    +{
        command => 'append',
        summary => 'Append data by KEY VALUE',
        description => <<'EODESC',
Description:
    Append VALUE after existing data which has specified KEY.

Usage:
    > append <KEY> <VALUE>
    > append mykey MySuffix
EODESC
    },
    +{
        command => 'prepend',
        summary => 'Prepend data by KEY VALUE',
        description => <<'EODESC',
Description:
    Put VALUE before existing data which has specified KEY.

Usage:
    > prepend <KEY> <VALUE>
    > prepend mykey MyPrefix
EODESC
    },
    +{
        command => 'cas',
        summary => 'Set data by KEY VALUE with CAS',
        description => <<'EODESC',
Description:
    Store data as specified (KEY, VALUE) if CAS of data is not modified.

Usage:
    > cas <KEY> <VALUE> <CAS> [<EXPIRE> [<FLAGS>]]
    > cas mykey1 MyValue1 1
    > cas mykey2 MyValue2 26  0     # Never expires. Default
    > cas mykey3 MyValue3 355 120 1
EODESC
    },
    +{
        command => 'touch',
        summary => 'Update data with new EXPIRE',
        description => <<'EODESC',
Usage:
    > touch <KEY> <EXPIRE>
    > touch foo 300  # will expire in 5 min
    > touch bar 0    # will never expire
EODESC
    },
    +{
        command => 'incr',
        summary => 'Add numeric VALUE for a data',
        description => <<'EODESC',
Usage:
    > incr <KEY> <VALUE>
    > incr foo 1
EODESC
    },
    +{
        command => 'decr',
        summary => 'Reduce numeric VALUE from a data',
        description => <<'EODESC',
Usage:
    > decr <KEY> <VALUE>
    > decr foo 1
EODESC
    },
    +{
        command => 'delete',
        summary => 'Delete data by KEY',
        description => <<'EODESC',
Usage:
    > delete <KEY>
EODESC
    },
    +{
        command => 'flush_all',
        summary => 'Invalidate whole data',
        description => <<'EODESC',
Usage:
    > flush_all [<DELAY>]
    > flush_all           # Invalidate immediately
    > flush_all 60        # Invalidate after 60 seconds
EODESC
    },
    +{
        command => 'call',
        summary => 'Execute any command',
        description => <<'EODESC',
Description:
    This command executes any commands given as arguments.
Usage:
    > call version
    > call get myKey
    > call stats conns
EODESC
    },
);

1;
__END__

=encoding utf-8

=head1 NAME

App::Memcached::CLI::Help - Utility to show commands help

=head1 SYNOPSIS

    use App::Memcached::CLI::Help;
    my @command_info = @App::Memcached::CLI::Help::COMMANDS_INFO;

=head1 DESCRIPTION

This package provides utility for commands help.

=head1 LICENSE

Copyright (C) IKEDA Kiyoshi.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

IKEDA Kiyoshi E<lt>progrhyme@gmail.comE<gt>

=cut

